go/state/state_db_test.go

```diff
@@ -4143,6 +4144,177 @@ func TestSlotIdOrder(t *testing.T) {
+	if got, want := db.GetTransientState(address1, key1), val2; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+}
+
+func TestStateDB_SetTransientState_ValueIsNotSetIfSame(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := CreateStateDBUsing(mock)
+
+	db.SetTransientState(address1, key1, val1)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	db.SetTransientState(address1, key1, val2)
+	if got, want := db.GetTransientState(address1, key1), val2; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+}
+
+func TestStateDB_SetTransientState_SettingSameValueReturnsEarly(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := createStateDBWith(mock, defaultStoredDataCacheSize, true)
+
+	db.SetTransientState(address1, key1, val1)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	if got, want := len(db.undo), 1; got != want {
+		t.Errorf("unexpected undo len, wanted: 1, got: %v", len(db.undo))
+	}
+
+	db.SetTransientState(address1, key1, val1)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	if got, want := len(db.undo), 1; got != want {
+		t.Errorf("unexpected undo len, wanted: 1, got: %v", len(db.undo))
+	}
+}
+
+func TestStateDB_SetTransientState_RollbackWithMultipleSteps(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := CreateStateDBUsing(mock)
+
+	s1 := db.Snapshot()
+
+	db.SetTransientState(address1, key1, val1)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	s2 := db.Snapshot()
+
+	db.SetTransientState(address1, key1, val2)
+	if got, want := db.GetTransientState(address1, key1), val2; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	db.RevertToSnapshot(s2)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	db.RevertToSnapshot(s1)
+	if got, want := db.GetTransientState(address1, key1), valEmpty; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+}
+
+func TestNonCommittableStateDB_TransientStorage_DoesNotLeakBetweenInstances(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := CreateNonCommittableStateDBUsing(mock)
+
+	db.SetTransientState(address1, key1, val1)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	db.Release()
+	// TransientStorage must be cleared
+	db2 := CreateNonCommittableStateDBUsing(mock)
```

Comment:
It is not guaranteed that `CreateNonCommittableStateDBUsing` is returning the StateDB instance released in line 4310 -- and it is increasingly unlikely when running lots of tests, since the pool of StateDB instances grows.

You can try and call `CreateNonCommittableStateDBUsing` multiple times without releaseing the DB you got until you retrieve the one you had.

Alternatively, the test could also call `resetState` explicitly and check that the transient state got reset.

Reply:
I guess explicitly calling `resetState` would be the best practice for this test.